### DISCUSSION TITLE
Feature Request: Add a wait timeout for GitHub Actions ephemeral runners to pick up a job.

### DISCUSSION THREAD

**Discussion by fduser1** [community/community#58215] [Actions]
### Select Topic Area

Product Feedback

### Body

Currently we are using ephemeral runners to pick up a single job and process it. We launch our runners on job triggers and expect them to immediately pick up the job and terminate after (ephemeral style). Because things happen, we have situations where our ephemeral runner instances wait/run indefinitely because there are no jobs to pick up (so much for ephemeral!). Such instances are burning our compute resources.

As part of configuring the GitHub Actions runner (e.g. ./config.sh), I would like to request a max wait timeout parameter that will be designated for picking up a job alone. If a job is not picked up by the ephemeral runner within the time threshold, then it should terminate. (The pick-up time should be independent of the time the runner needs to perform any optional self-update ops during load time.)

**Comment by triple3567** [comment]
Bump. This is a feature we would like as well.

**Comment by e10harvey** [comment]
We would like this feature too.

**Comment by ric-whitehead-bbc** [comment]
Bump

**Comment by e10harvey** [comment]
I did triage this and, by trial and error, discovered that the following can be done as of self-hosted runner version `2.316.0`:

1. Set a timeout in the runner client. One must start `run.sh` in the background and then start a timer since `run.sh` is blocking.
2. If the timeout is reached, attempt to de-register the runner via:
```
/path/to/config.sh remove ...
```
3. Check the return code of `/path/to/config.sh remove`. If it is `0`, exit with success. If it is non-zero, do not exit.

The following changes would be preferred:
1. Documentation of the above `config.sh` behavior, with examples.
2. Well documented public APIs in addition to `config.sh`, `run.sh`, etc.
3. A timeout option for `config.sh`.

**Reply by spamse** [reply]
I'm not sure this is what is being asked for, though I could be misunderstanding. The use-case is where the Runner applications has launched and is connected to the repository, awaiting a Job. If no Job picked up within the timeout period, then quit the application (at which point the normal ephemeral deregistration behaviour would occur).

**Comment by pauldraper** [comment]
Note that doing this in the runner executable itself is the only reliable mechanism.

You can kludge something together with pre script and post script and background processes, but it still is subject to race conditions where the runner receives a job as the timeout passes.

This is not limited to ephemeral runners; this a problem with scaling down any kind of runner.

**Reply by e10harvey** [reply]
In general I agree, but, you can work around the data race since config.sh tells you whether the server unregistered the ephemeral runner with either a 0 or non-zero return code. This isn't documented anywhere as far as I can tell though.

**Comment by breathe** [comment]
We would very much like this feature as well.  the `--ephemeral` option is very nice but its missing the ability to reliably shut down idle runners that don't ever receive a job for _some reason_.  

In our environment we spin up ephemeral runners on demand in response to the github workflow_job webhook and in some cases we use highly specific runner labels (which might include highly workflow specific information like commit id) to ensure that the runner in which the job gets executed precisely matches the originating github action workflow_job event.  If the workflow gets cancelled these runners can sit around idly until they reach an execution timeout.

We would like them to instead shutdown and unregister if they don't receive a job within ~a minute after being spun up.

**Reply by breathe** [reply]
To further motivate this (and to save somebody else a couple of hours) -- here's the super kludge i've just implemented with some hacky bash logic wrapping around runner's `run.sh` entrypoint ...  

```bash
#!/bin/bash

...

# configure the runner to run a script in response to picking up a job
export ACTIONS_RUNNER_HOOK_JOB_STARTED=/app/gha/run-before-job-starts.sh

echo running: /app/actions-runner/config.sh --unattended --name "runner-$JOB_NAME-$RAND_ID" --url https://github.com/zephr-xyz/zephyrus --token $GITHUB_ACTION_RUNNER_TOKEN --ephemeral --disableupdate --no-default-labels --labels "$LABELS"
/app/actions-runner/config.sh --unattended --name "runner-$JOB_NAME-$RAND_ID" --url https://github.com/zephr-xyz/zephyrus --token $GITHUB_ACTION_RUNNER_TOKEN --ephemeral --disableupdate --no-default-labels --labels "$LABELS"

# create a file which will be used to detect if the runner is idle
# the ACTIONS_RUNNER_HOOK_JOB_STARTED script will remove the file when job starts which will indicate to the code below that the runner is not idle
touch /app/gha/idle_runner_if_this_file_exists

# forward signals to the runner process
forward_signal() {
  local signal=$1
  echo "Forwarding signal $signal to child process $RUNNER_PID"
  kill -"$signal" "$RUNNER_PID"
}

trap 'forward_signal SIGHUP' SIGHUP
trap 'forward_signal SIGINT' SIGINT
trap 'forward_signal SIGTERM' SIGTERM
trap 'forward_signal SIGQUIT' SIGQUIT

# Invoke the runner script in the background and capture its PID
/app/actions-runner/run.sh &

RUNNER_PID=$!

# wait for either (1) 60 seconds or (2) run.sh to return -- which happens first

START_TIME=$(date +%s)

while true; do
  # Check if the runner process is still running
  if ! kill -0 $RUNNER_PID 2>/dev/null; then
    echo "runner has completed"
    break
  fi

  # Calculate the elapsed time
  CURRENT_TIME=$(date +%s)
  ELAPSED_TIME=$((CURRENT_TIME - START_TIME))

  # Check if 60 seconds have passed
  if [ $ELAPSED_TIME -ge 60 ]; then
    break
  fi

  # Sleep for a short interval before checking again
  sleep 1
done

# at this point, either run.sh has returned or 60 seconds have passed ...

if [ -f /app/gha/idle_runner_if_this_file_exists ]; then
  # Check if this file exists and if it does then signal the runner to unregister and shutdown because its been idle for 60 seconds
  # this exists to handle edge case that can occur when a gha is cancelled
  # -> github signals the need to spin up a worker
  # -> worker launched and starts initializing
  # -> github cancels the workflow prior to the worker starting the job
  # -> worker registers with github and requests a job but there is nothing to do since the job was cancelled
  # -> worker sits idle for 60 seconds and therefore /app/gha/idle_runner_if_this_file_exists still exists so the worker gets shutdown
  # if this logic fails and the worker remains idle -- the worker will sit there idle until it hits the cloud job deadline (which we have set to an hour currently) -- or until a
  # matching job becomes available.
  echo "$(date +"%Y-%m-%dT%H:%M:%S%z") runner has been idle for 60 seconds -- killing it"
  kill $RUNNER_PID
  echo "$(date +"%Y-%m-%dT%H:%M:%S%z") signal sent to worker to stop after idling for 60 seconds"
else
  echo "$(date +"%Y-%m-%dT%H:%M:%S%z") runner is not idle after 60 seconds -- allowing it to continue to completion"
fi

echo "$(date +"%Y-%m-%dT%H:%M:%S%z") waiting for worker to complete"

# Wait for the runner process to finish before exiting with same process code as runner
wait $RUNNER_PID
exit $?
```

with /app/gha/run-before-job-starts.sh
```bash
#!/bin/bash

# This script is invoked when a runner receives a job
# we remove idle_runner_if_this_file_exists file to indicate that the runner got a job and is not idle
# This tells logic within start-github-action-runner-linux.sh not to kill the runner ...
rm /app/gha/idle_runner_if_this_file_exists
echo "$(date +"%Y-%m-%dT%H:%M:%S%z") picked up a job -- canary file idle_runner_if_this_file_exists removed"
```

As mentioned by pauldraper above

> still is subject to race conditions where the runner receives a job as the timeout passes

That comment applies to this implementation -- however if the worker was sitting for ~60 seconds trying to register with github/pick up a job and then remove one file then something's likely already gone pretty wrong ... cancelling such a fudged up worker not likely significantly worse than just letting it continue to be potentially fudged -- and coming down on the wrong side of the race will cancel the job unecessarily 60 seconds after it was supposed to have started doing work but was instead sitting idle -- which shouldn't be a hugely costly failure mode ...  

normally from what i've observed so far -- if a job is available when the worker registers then getting the job should take less than a second and then removing the file should happen quite quickly after that so I'm hoping the race is never lost except when there are github action service outage type scenarios in effect anyway ...
